ls1intum/Artemis
# Feature/programming exercise/mock connector requests for tests

### Checklist

### Motivation and Context


Currently, we always mock our connector services (e.g. Bitbucket, Bamboo) when testing programming exercises. We want to only mock the requests to the CI and VCS servers and not the Spring service itself, so we also cover the code in these cases. 

### Description

I added two new tests (setup and start programming exercise), for  which all requests to external services get mocked by Spring's `MockRestServiceServer`. The added benefit, besides that we actually run the code in our `BambooService` and `BitbucketService` is that we also test the actual values that requests to the external services contain. A request gets mocked by matching the contents and headers against some predefined values. Only when these values are equal, the request gets mocked. Otherwise, an exception is thrown. Because of that, the test will only succeed if the connector services process the parameters as expected.

**Because this PR would be too big, I made a second PR containing only the production code changes: #1169 . After we merge this, I will finalize this PR.**
In order to be able to mock everything, I also had to make a couple of other adaptions:
- I added a `BambooServerConfiguration` that creates beans for the Bamboo server and client. Before that, these objects got instantiated in the methods themselves, so we couldn't mock anything that got routed through them. Since we send requests through these objects, we also have to be able to inject mock instances, so I had to refactor the configuration into Spring beans.
- The `BitbucketBambooUpdateService` is not new! I just refactored the old `BambooUpdateService` that only encapsulated the `BitbucketBambooUpdateService` in an inner class. This made no sense and meant only having one additional unused object.
- The Bitbucket and Bamboo services now have dedicated `RestTemplates` that are necessary for mocking the REST requests for both of them separately . Those templates still are plain Spring `RestTemplates` without any modification.
- The `BambooBuildPlanUpdateProvider` is **unmodified** code from the old `BitbucketBambooUpdateService` that @krusche once extracted from the third party `BambooClient`. I had to move this code to a separate bean, because it also calls services, that perform REST requests (but not using Spring's RestTemplate). That is why we have to mock this whole service when executing tests, which is only possible if it is an injected bean.
- There is a couple of new DTOs that we send to the Bitbucket/Bamboo servers. In our old code,  we just manually put those values into Maps and serialized them. Testing this is not ideal, because we would have to copy a lot of code. This way, all models get managed in those DTOs and we can be sure, that both the tests and the productive code use the same data structures (So testing for equality is easier).

In this PR:
- I also started moving String values for our REST API endpoints to constant classes, so we don't have to copy them in our tests, but can just reference the constant, so we can guarantee that the productive and the test code both use the same values.

### Steps for Testing


1. Run all Java tests
2. Create, edit and participate in a programming exercise. Everything should work as before.
